src/modules/rlm_python/rlm_python.c

```diff
@@ -488,95 +488,147 @@ static int mod_populate_vptuple(rlm_python_t const *inst, request_t *request, Py
-	}
+	for (vp = fr_cursor_init(&cursor, &vps); vp; vp = fr_cursor_next(&cursor)) tuplelen++;
 
-	if (tuple_len == 0) {
-		Py_INCREF(Py_None);
-		p_arg = Py_None;
-	} else {
-		int i = 0;
-		if ((p_arg = PyTuple_New(tuple_len)) == NULL) {
-			rcode = RLM_MODULE_FAIL;
-			goto finish;
+	if ((vps_tuple = PyTuple_New(tuplelen)) == NULL) goto error;
+
+	for (vp = fr_cursor_init(&cursor, &vps); vp; vp = fr_cursor_next(&cursor), i++) {
+		PyObject *pPair = NULL;
+
+		/* The inside tuple has two only: */
+		if ((pPair = PyTuple_New(2)) == NULL) goto error;
+
+		if (mod_populate_vptuple(inst, request, pPair, vp) == 0) {
+			/* Put the tuple inside the container */
+			PyTuple_SET_ITEM(vps_tuple, i, pPair);
+		} else {
+			Py_DECREF(pPair);
+			goto error;
 		}
+	}
+	PyTuple_SET_ITEM(p_arg, pos, vps_tuple);
+	return true;
 
-		for (vp = fr_cursor_init(&cursor, &request->request_pairs);
-		     vp;
-		     vp = fr_cursor_next(&cursor), i++) {
-			PyObject *pp;
+error:
+	Py_XDECREF(vps_tuple);
+	return false;
+}
 
-			/* The inside tuple has two only: */
-			if ((pp = PyTuple_New(2)) == NULL) {
-				rcode = RLM_MODULE_FAIL;
-				goto finish;
-			}
+static unlang_action_t do_python_single(rlm_rcode_t *p_result,
+					rlm_python_t const *inst, request_t *request, PyObject *p_func, char const *funcname)
+{
+	PyObject	*p_ret = NULL;
+	PyObject	*p_arg = NULL;
+	PyObject 	*p_dict = NULL;
+	PyObject    *p_tuple_int = NULL;
+	PyObject    *update_dict = NULL;
+	rlm_rcode_t	rcode;
 
-			if (mod_populate_vptuple(inst, request, pp, vp) == 0) {
-				/* Put the tuple inside the container */
-				PyTuple_SET_ITEM(p_arg, i, pp);
-			} else {
-				Py_INCREF(Py_None);
-				PyTuple_SET_ITEM(p_arg, i, Py_None);
-				Py_DECREF(pp);
-			}
+	/* Default return value is "OK, continue" */
+	rcode = RLM_MODULE_OK;
+
+	/*
+	 * p_arg is a 4-tuple with (Request, Reply, Control and State)
+	 * If some list is not available, NONE is used instead
+	 */
+	p_arg = PyTuple_New(4);
+	if (!p_arg) {
+		ERROR("%s - Memory cannot be allocated for PyTyple_New", funcname);
+		rcode = RLM_MODULE_FAIL;
+		goto finish;
+	}
+
+	/* If there is a request, fill in the first 4 attribute lists */
+	if (!request) {
+		int i = 0;
+
+		/* If there is no request, set all the elements to None */
+		for (i = 0; i < 4; i++) mod_populate_vps(inst, request, p_arg, i, NULL);
+	} else {
+		if (!mod_populate_vps(inst, request, p_arg, 0, request->request_pairs) ||
+		    !mod_populate_vps(inst, request, p_arg, 1, request->reply_pairs) ||
+		    !mod_populate_vps(inst, request, p_arg, 2, request->control_pairs) ||
+		    !mod_populate_vps(inst, request, p_arg, 3, request->state_pairs))
+		{
+
+			ERROR("%s - mod_populate_vps failed", funcname);
+			rcode = RLM_MODULE_FAIL;
+			goto finish;
 		}
 	}
 
 	/* Call Python function. */
-	p_ret = PyObject_CallFunctionObjArgs(p_func, p_arg, NULL);
+	p_dict = PyDict_New();
+	if (!p_dict ||
+	    PyDict_SetItemString(p_dict, "request", PyTuple_GET_ITEM(p_arg, 0)) ||
```

Comment:
Just move this name setting to `mod_populate_vps()`, and pass the name to that function.  less code == better code